File: khelmel-java/src/main/java/io/wulfcodes/khelmel/model/bingo/BingoRoom.java
package io.wulfcodes.khelmel.model.bingo;

import io.javalin.websocket.WsContext;

import java.util.ArrayList;
import java.util.List;

public class BingoRoom {

    private final String roomId;
    private final List<BingoPlayer> players = new ArrayList<>(2);
    private GameState state = GameState.WAITING;
    private String currentTurn; // username of player whose turn it is

    public BingoRoom(String roomId) {
        this.roomId = roomId;
    }

    public String getRoomId() { return roomId; }
    public GameState getState() { return state; }
    public void setState(GameState state) { this.state = state; }
    public String getCurrentTurn() { return currentTurn; }
    public void setCurrentTurn(String currentTurn) { this.currentTurn = currentTurn; }

    public boolean isFull() { return players.size() >= 2; }
    public int getPlayerCount() { return players.size(); }

    /** Returns true if player was added, false if room is full or username taken. */
    public synchronized boolean addPlayer(String username, WsContext ctx) {
        if (isFull()) return false;
        if (players.stream().anyMatch(p -> p.getUsername().equals(username))) return false;
        players.add(new BingoPlayer(username, ctx));
        if (isFull()) state = GameState.LOBBY;
        return true;
    }

    public synchronized void removePlayer(String username) {
        players.removeIf(p -> p.getUsername().equals(username));
        if (players.size() < 2 && state != GameState.FINISHED) {
            state = GameState.WAITING;
        }
    }

    public BingoPlayer getPlayer(String username) {
        return players.stream()
            .filter(p -> p.getUsername().equals(username))
            .findFirst().orElse(null);
    }

    public BingoPlayer getOpponent(String username) {
        return players.stream()
            .filter(p -> !p.getUsername().equals(username))
            .findFirst().orElse(null);
    }

    public List<BingoPlayer> getPlayers() { return players; }

    public int getReadyCount() {
        return (int) players.stream().filter(BingoPlayer::isReady).count();
    }

    /** Broadcast a JSON string to all players in the room. */
    public void broadcast(String message) {
        for (BingoPlayer p : players) {
            try {
                if (p.getSession().session.isOpen()) {
                    p.getSession().send(message);
                }
            } catch (Exception ignored) {}
        }
    }

    /** Send a JSON string to a specific player. */
    public void sendTo(String username, String message) {
        BingoPlayer player = getPlayer(username);
        if (player != null) {
            try {
                if (player.getSession().session.isOpen()) {
                    player.getSession().send(message);
                }
            } catch (Exception ignored) {}
        }
    }
}

File: khelmel-js/src/models/bingo.room.js
const BingoPlayer = require('./bingo.player');
const GameState = require('./game.state');

class BingoRoom {
    constructor(roomId) {
        this.roomId = roomId;
        this.players = [];
        this.state = GameState.WAITING;
        this.currentTurn = null; // username
    }

    getRoomId() { return this.roomId; }
    getState() { return this.state; }
    setState(state) { this.state = state; }
    getCurrentTurn() { return this.currentTurn; }
    setCurrentTurn(currentTurn) { this.currentTurn = currentTurn; }

    isFull() { return this.players.length >= 2; }
    getPlayerCount() { return this.players.length; }

    /** Returns true if player was added, false if room is full or username taken. */
    addPlayer(username, ws) {
        if (this.isFull()) return false;
        if (this.players.some(p => p.getUsername() === username)) return false;
        this.players.push(new BingoPlayer(username, ws));
        if (this.isFull()) this.state = GameState.LOBBY;
        return true;
    }

    removePlayer(username) {
        this.players = this.players.filter(p => p.getUsername() !== username);
        if (this.players.length < 2 && this.state !== GameState.FINISHED) {
            this.state = GameState.WAITING;
        }
    }

    getPlayer(username) {
        return this.players.find(p => p.getUsername() === username) || null;
    }

    getOpponent(username) {
        return this.players.find(p => p.getUsername() !== username) || null;
    }

    getPlayers() { return this.players; }

    getReadyCount() {
        return this.players.filter(p => p.isReady()).length;
    }

    broadcast(message) {
        const payload = JSON.stringify(message);
        for (const p of this.players) {
            try {
                if (p.getSession().readyState === 1) { // OPEN
                    p.getSession().send(payload);
                }
            } catch (e) { }
        }
    }

    sendTo(username, message) {
        const player = this.getPlayer(username);
        if (player) {
            try {
                if (player.getSession().readyState === 1) { // OPEN
                    player.getSession().send(JSON.stringify(message));
                }
            } catch (e) { }
        }
    }
}

module.exports = BingoRoom;
